src/core/toga/widgets/optioncontainer.py

```diff
@@ -3,6 +3,63 @@
 from .base import Widget
 
 
+class OptionItem:
+    def __init__(self, container, label, widget, enabled):
+        self._container = container
+        self._index = len(container)
+        self._label = label
+        self._widget = widget
+        self._enabled = enabled
```

Comment:
Rather than having a duplicate of this data, it would be preferable to have it read *directly* from the actual widget. Is it possible to read the enabled/disable state from the widget itself?

Reply:
The problem I'm facing here is that I'm not saving OptionItem Widget `_impl` anywhere.  This `OptionItem` class is only saving  the widget (`self._widget`) inside the OptionItem.

So, to access to `label` and `enabled` values I have to create some methods in the implementation layer like for example `get_label` and `is_enabled` and at the moment I only have `OptionContainer` implementation class to make this methods.

But rather than create that methods in the `OptionContainer`, It would be better have an `OptionItem` in the implementation layer, to put all the specific `OptionItem` functionality detached of the `OptionContainer`. Do you think this is the best way to do it?

Reply:
I'm not sure I follow. The OptionItem has a link to the OptionContainer interface, from which you can get to the OptionContainer impl; If the impl layer has an `set_option_enabled(index, state)` and `is_option_enabled(index)`, that would allow the interface to directly access the state of the widget by providing the context that is needed (i.e., the index/label) to answer the request. 

The key insight here is that the API at the *impl* layer is purely functional - it's whatever lets us get the job done. It doesn't need to match the public interface-layer API. We can have methods on the impl layer that let us achieve functional goals, but don't have any directly corresponding method on the interface layer. That's one of the benefits of a layered API - we can expose a completely clean API to the end user, but internally use whatever API makes sense to bridge the differences between platforms.

We're already doing a similar thing with Tree/Table widgets. The public API for the Tree/Table is based around data sources; however, the impl layer talks about adding rows, refreshing rows, etc.

Does that make sense?

Reply:
> The key insight here is that the API at the impl layer is purely functional
I based on this to make changes, I wasn't sure about what do or not to do in impl layer but now I think it's clear. Thanks!